### PR TITLE
[feature](schema change) unified schema change for parquet and orc reader

### PR DESCRIPTION
**Pull request by AshinGau**
## Proposed changes

Following https://github.com/apache/doris/pull/25138, unified schema change interface for parquet and orc reader, and can be applied to other format readers as well.

## Supported Type Changes
More type changes are supported:
image

## ColumnTypeConverter
Unified schema change interface for all format readers:
- First, read the data according to the column type of the file into source column;
- Second, convert source column to the destination column with type planned by FE

## PhysicalToLogicalConverter
Convert parquet physical column to logical column
In parquet document(https://github.com/apache/parquet-format/blob/master/LogicalTypes.md),
Logical or converted type is the data type of column, physical type is the stored type of column chunk. eg, decimal type can be stored as INT32, INT64, BYTE_ARRAY, FIXED_LENGTH_BYTE_ARRAY, so there is a convert process from physical type to logical type. In addition, schema change will bring about a change in logical type.

In previous implementations, physical and logical conversion were mixed together, resulting in severe code complexity and bloating.
`PhysicalToLogicalConverter` strips away the conversion of logical type, and reuse `ColumnTypeConverter` to resolve schema change, allowing parquet reader to only focus on the conversion of physical types.

Therefore, tow layers converters are designed:
- First, read parquet data with the physical type
- Second, convert physical type to logical type
- Third, convert logical type to the final type planned by FE(schema change)

### Ultimate performance optimization:
1. If process of (First => Second) is consistent, eg. from BYTE_ARRAY to string, no additional copies and conversions will be introduced;
2. If process of (Second => Third) is consistent, eg. from decimal(12, 4) to decimal(8, 2), no additional copies and conversions will be introduced;
3. Null map is share among all processes, no additional copies and conversions will be introduced in null map;
4. Only create one physical column in physical conversion, and reused in each loop;
5. Only create one logical column in logical conversion, and reused in each loop;
6. FIXED_LENGTH_BYTE_ARRAY is read as ColumnUInt8 instead of ColumnString, so the underlying decoder has no process to decode string and use memory copy to read the data as a whole, and the conversion has no need to resolve the Offsets in ColumnString.

### Remaining Issues
1. Not support paimon timestamp type, because it has no logical or converted type, and is conflict with column type change from bigint to timestamp. Reference paimon issue: https://github.com/apache/paimon/pull/2739, 

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/exec/format/column_type_convert.cpp [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    // change to string type
+    // example: decimal -> string
+    if (is_string_type(dst_primitive_type)) {
+        return _to_string_converter(src_type, dst_type);
+    }
+
+    // string type to other type
+    // example: string -> date
+    if (is_string_type(src_primitive_type)) {
+        return _cast_string_converter(src_type, dst_type);
```
Comment (morningman) [inline be/src/vec/exec/format/column_type_convert.cpp:305]:
```suggestion
        return _from_string_converter(src_type, dst_type);
```

File: be/src/vec/data_types/data_type_struct.cpp [hunk 2]
```diff
@@ -355,14 +355,7 @@ std::optional<size_t> DataTypeStruct::try_get_position_by_name(const String& nam
 }
 
 String DataTypeStruct::get_name_by_position(size_t i) const {
-    if (i == 0 || i > names.size()) {
```
Comment (morningman) [inline be/src/vec/data_types/data_type_struct.cpp:358]: Why removing this check?

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 3]
```diff
@@ -532,6 +532,14 @@ std::vector<OrcPredicate> value_range_to_predicate(
         std::vector<orc::TypeKind>* unsupported_pushdown_types) {
     std::vector<OrcPredicate> predicates;
 
+    PrimitiveType src_type = OrcReader::convert_to_doris_type(type).type;
+    if (src_type != primitive_type) {
+        if (!(is_string_type(src_type) && is_string_type(primitive_type))) {
```
Comment (morningman) [inline be/src/vec/exec/format/orc/vorc_reader.cpp:537]: why?

File: be/src/vec/exec/format/parquet/parquet_column_convert.h [hunk 4]
```diff
@@ -193,195 +106,189 @@ struct ConvertParams {
...
-    }
+/**
+ * Convert parquet physical column to logical column
+ * In parquet document(https://github.com/apache/parquet-format/blob/master/LogicalTypes.md),
+ * Logical or converted type is the data type of column, physical type is the stored type of column chunk.
+ * eg, decimal type can be stored as INT32, INT64, BYTE_ARRAY, FIXED_LENGTH_BYTE_ARRAY.
+ * So there is a convert process from physical type to logical type.
+ * In addition, Schema change will bring about a change in logical type.
+ *
+ * In previous implementations, physical and logical conversion were mixed together,
```
Comment (morningman) [inline be/src/vec/exec/format/parquet/parquet_column_convert.h:137]: No need to add this in comment

File: be/src/vec/exec/format/column_type_convert.h [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+    }
+};
+
+template <>
+struct SafeCastString<TYPE_LARGEINT> {
+    static bool safe_cast_string(
+            const char* startptr, const int buffer_size,
+            PrimitiveTypeTraits<TYPE_LARGEINT>::ColumnType::value_type* value) {
+        int64 cast_to_int = 0;
+        bool can_cast = safe_strto64(startptr, buffer_size, &cast_to_int);
```
[the remark this answers is not in the document]
Comment (AshinGau) [reply be/src/vec/exec/format/column_type_convert.h:284]: Use `read_int_text_impl` instead.

File: be/src/vec/data_types/data_type_struct.cpp (hunk 2 above)
Comment (AshinGau) [reply be/src/vec/data_types/data_type_struct.cpp:358]: This function is wrong, and there's no check like `Block::get_by_position`

File: be/src/vec/exec/format/orc/vorc_reader.cpp (hunk 3 above)
Comment (AshinGau) [reply be/src/vec/exec/format/orc/vorc_reader.cpp:537]: Altering column type from `string` to `varchar` can still use push-down predicates.

**Comment by AshinGau** [comment]
run buildall

**Review by morningman — DISMISSED** [review]

**Comment by AshinGau** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by AshinGau** [comment]
run buildall

**Comment by AshinGau** [comment]
run buildall

**Review by morningman — APPROVED** [review]

**Review by kaka11chen — APPROVED** [review]
LGTM

**Pull request merged**